dataStructures/queues/queue.go

```diff
@@ -45,25 +45,42 @@ func (q *Queue) Add(order *Order) {
 		*q = Queue{order}
 	} else {
 		var appended bool
+		var insert int
+
+		// using binary-search to know index to insert the order.
+		lowIndex := 0
+		highIndex := len(*q) - 1
 
 	QLoop:
-		for i, existingOrder := range *q {
-			var insert int
+		for lowIndex <= highIndex {
+
+			midPoint := (lowIndex + highIndex) / 2
+			guessOrder := (*q)[midPoint]
 
 			switch {
-			case order.priority < existingOrder.priority:
-				insert = i
-			case order.priority == existingOrder.priority:
-				insert = 1 + i
-				// peek to check next order in the queue.
-				if insert < len(*q) && (*q)[insert].priority == order.priority {
-					// keep looking for the best location to insert.
-					continue QLoop
-				}
+			case order.priority < guessOrder.priority:
+				// we move closer to the left-side
+				highIndex = midPoint - 1
+			case order.priority > guessOrder.priority:
+				// we move to the right-side of the queue
+				lowIndex = midPoint + 1
 			default:
-				continue QLoop
+				// when the priority is equal
+				// peek to the next order to check the priority
+				nextIndex := midPoint + 1
+				if nextIndex < len(*q) && order.priority == (*q)[nextIndex].priority {
+					// found the insertion point
+					insert = nextIndex
+					appended = true
+					break QLoop
+				}
+				lowIndex = midPoint + 1
```

Comment:
Let's remove the `default` case entirely so we explicitly account for all cases. The "found" case is now `order.priority == guessOrder.priority || lowIndex == highIndex`. The first means we've found the right part of the queue because we match something else, and the second means we're at least in between two other Orders with the "sandwiching" priorities.

I'd recommend removing the `insert = ` and `appended = ` declarations, and just insert the Order in the switch case.